File: backend/model/message_model.py
from typing import Literal, Optional
from pydantic import BaseModel, Field
from pydantic import ConfigDict
from model.object_id import PyObjectId
from datetime import datetime

class MessageModel(BaseModel):
    id: PyObjectId = Field(default_factory=PyObjectId, alias="_id")
    conv_id: str
    ts: int = Field(default_factory=lambda: int(datetime.now().timestamp()))
    sender: Literal["ai", "user"]
    message: str
    correction: Optional[str] = None
    token_count: Optional[int] = None
    error: Optional[bool] = None

    model_config = ConfigDict(
        populate_by_name=True,
        arbitrary_types_allowed=True,
        extra='ignore'
    )

File: mobile/penpal/penpal/Messages/MessageModel.swift
//
//  MessageModel.swift
//  penpal
//

struct MessageModel: Codable, Identifiable {
    var id: String
    var message: String
    var sender: String
    var ts: Double
    var correction: String?
    var error: Bool?

    init(id: String, message: String, sender: String, ts: Double, correction: String? = nil, error: Bool? = nil) {
        self.id = id
        self.message = message
        self.sender = sender
        self.ts = ts
        self.correction = correction
        self.error = error
    }
}
